### ISSUE TITLE
Bad continuation with trailing space in the prompt

### ISSUE THREAD

**Issue by xefoci7612** [karpathy/llama2.c#180]
Here is an example of a continuation with dot instead of a better one like "_and_":
```
./run /tmp/ramdisk/model.bin 0 20 "One day, Lily met a Shoggoth "
<s>
One day, Lily met a Shoggoth . He was very excited and wanted to show
```

Without trailing space everything is ok:
```
./run /tmp/ramdisk/model.bin 0 20 "One day, Lily met a Shoggoth"
<s>
One day, Lily met a Shoggoth. He was very excited and wanted to show it
```

I looked at the emitted tokens:
```
Processing 1 <
<s>
>
Processing 6716 <One>
Processing 2462 < day>
Processing 47 <,>
Processing 28624 < Lil>
Processing 124 <y>
Processing 1539 < met>
Processing 263 < a>
Processing 17550 < Sho>
Processing 1505 <gg>
Processing 720 <oth>
Processing 35 < >           <----- Here prompt ends with a space
Processing 29889 <.>        <----- Here model continues with a dot
Processing 940 < He>
Processing 471 < was>
Processing 1407 < very>
Processing 24173 < excited>
Processing 322 < and>
Processing 5131 < wanted>
Processing 304 < to>
```

Not clear the root cause of this, maybe due to the fact that Llama tokens happen have the space prepended instead of appended.

I have looked up in the vocab and both tokens for " and" (322) and for "and" (392) exist. The appended space version "and " does not exist.

My guess is that sentecepiece tokenizes usually with the space prepended, so that during training LLama model sees very few tokens corresponding to "and" and many more tokens of " and".

Here is yet another example among many. It seems that fails almost always when prompt ends with a trailing space.

```
$ ./run /tmp/ramdisk/model.bin 0 20 "One day, Lily met a Shoggoth and"
<s>
One day, Lily met a Shoggoth and a little girl. The girl was very excited
achieved tok/s: 279.411765
$ ./run /tmp/ramdisk/model.bin 0 20 "One day, Lily met a Shoggoth and "
<s>
One day, Lily met a Shoggoth and wel. They were both very excited and
achieved tok/s: 275.362319

```

**Comment by kroggen** [comment]
Yes. The application must remove trailing spaces from the prompt.

**Comment by karpathy** [comment]
This is a more general issue that goes well beyond spaces and more generally the issue is with prompts that end "mid-token". The example I give on the readme correct has no trailing space. I could be in favor of, just like the OpenAI API, warning if a prompt ends in a space, but deleting in app I'm not sure is the right thing.

**Comment by xefoci7612** [comment]
@karpathy thank you now I understand.

Indeed removing the trailing space would be a half-cooked hack. IMHO a possible **real** solution could be the following:

1) Given a prompt take the _last_token_

2) Then build the _token_candidates_ set looping across the vocab and picking all the tokens whose corresponding string **starts with** _last_token_'s string. Note that _last_token_ will be always included in _token_candidates_

3) Process the prompt with _last_token_ removed

4) At the first step after the prompt is processed, before sampling the next token, pick one only among the _token_candidates_ set. For instance, after softmax,, zero the score in _state.logits_ of the tokens that are not in _token_candidates_

I can write a simple patch if you are interested.

**Comment by xefoci7612** [comment]
@karpathy  here is my attempt at fixing it in a general way #191
